### PR TITLE
[DebugInfo][RemoveDIs] Extract DPValues in CodeExtractor like dbg.values

### PR DESCRIPTION
**Pull request by jmorse**
CodeExtractor shifts dbg.value intrinsics out of the region being extracted and updates them to be appropriate in the extracted function. With new non-intrinsic variable locations, we need to manually do this too, with DPValues.

Most of this patch shifts and refactors some utilities in fixupDebugInfoPostExtraction so that we can add a single extra helper lambda that iterates over DPValues and applies update-utilities. We also have to assign the IsNewDbgInfoFormat flag in a bunch of places -- this normally gets set the moment you insert a block into a function (or function into a module), however a few blocks are constructed here before being inserted, thus we have to do some manual setup.

Tested via LoopExtractor_alloca.ll, which invokes debugify.

### CODE REVIEW

File: llvm/lib/Transforms/Utils/CodeExtractor.cpp [hunk 1]
```diff
@@ -1552,8 +1568,41 @@ static void fixupDebugInfoPostExtraction(Function &OldFunc, Function &NewFunc,
...
+          NewScope, OldVar->getName(), OldVar->getFile(), OldVar->getLine(),
+          OldVar->getType(), /*AlwaysPreserve=*/false, DINode::FlagZero,
+          OldVar->getAlignInBits());
+    }
+    return cast<DILocalVariable>(NewVar);
+  };
+
+  auto UpdateDPValuesOnInst = [&](Instruction &I) -> void {
+    if (!I.hasDbgValues())
+      return;
```
Comment (OCHyams) [inline llvm/lib/Transforms/Utils/CodeExtractor.cpp:1589]: Is this check really needed?

File: llvm/lib/Transforms/Utils/CodeExtractor.cpp [hunk 2]
```diff
@@ -1552,8 +1568,41 @@ static void fixupDebugInfoPostExtraction(Function &OldFunc, Function &NewFunc,
...
+    for (auto &DPV : I.getDbgValueRange()) {
+      // Apply the two updates that dbg.values get: invalid operands, and
+      // variable metadata fixup.
+      if (any_of(DPV.location_ops(), IsInvalidLocation)) {
+        DPVsToDelete.push_back(&DPV);
+        continue;
+      }
```
Comment (OCHyams) [inline llvm/lib/Transforms/Utils/CodeExtractor.cpp:1596]: `// FIXME: Support dbg.assign intrinsics.`

**Review by OCHyams — APPROVED** [review]
LGTM

File: llvm/lib/Transforms/Utils/CodeExtractor.cpp (hunk 1 above)
Comment (jmorse) [reply llvm/lib/Transforms/Utils/CodeExtractor.cpp:1589]: I suppose not -- moderately depends on whether there's a performance benefit. Which can be ignored for the moment.

**Comment by jmorse** [comment]
NB: not applying all the clang-format recommendations as it affects lines I'm not editing, and I don't want to pollute git-blame

**Pull request merged**